### PR TITLE
Enhancement 473 map store docker compose

### PR DESCRIPTION
**Pull request by agpenton**
### Changes:
* Changes on the syntax of the Dockerfile that resolves geosolutions-it/DevOps#471.
* Created the docker-compose file that resolves geosolutions-it/DevOps#473.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,22 +1,49 @@
...
+ENV MAXIMUM_MEMORY="512m"
+ENV JAVA_OPTS="${JAVA_OPTS}  -Xms${INITIAL_MEMORY} -Xmx${MAXIMUM_MEMORY} -XX:MaxPermSize=128m"
+ENV GEOSTORE_OVR_OPT="-Dgeostore-ovr=${CATALINA_BASE}/conf/${OVR} \
+                      -Duser.timezone=UTC"
+ARG TOMCAT_EXTRAS=false
+ARG OVR=""
 
 # Optionally remove Tomcat manager, docs, and examples
-ARG TOMCAT_EXTRAS=false
 RUN if [ "$TOMCAT_EXTRAS" = false ]; then \
```
Comment (randomorder) [inline Dockerfile:23]: I think the whole "remove TOMCAT_EXTRAS" stuff is not needed anymore with the new tomcat base image:
```
docker run -it tomcat:9-jdk11-openjdk bash
Unable to find image 'tomcat:9-jdk11-openjdk' locally
9-jdk11-openjdk: Pulling from library/tomcat
b9a857cbf04d: Already exists
d557ee20540b: Already exists
3b9ca4f00c2e: Already exists
667fd949ed93: Already exists
b6a945768a17: Already exists
aa2f45feafbf: Already exists
7179e0e40c20: Already exists
d659cf2d36fd: Already exists
880a7c791192: Pull complete
dd17efe07475: Pull complete
Digest: sha256:e7bc48bd0b0c18db84e718769ac27c87f90f9ecb20b761700be0a016a7f840d2
Status: Downloaded newer image for tomcat:9-jdk11-openjdk
root@8ba65a0295ff:/usr/local/tomcat# ls webapps

```

the webapps directory is empty already in the base image itself

File: Dockerfile [hunk 2]
```diff
@@ -1,22 +1,49 @@
...
+#COPY ./docker/wait-for-postgres.sh ${CATALINA_BASE}/
+# adding the env
+RUN if [ "${OVR}" = "geostore-datasource-ovr.properties" ]; then \
+        cp ${CATALINA_BASE}/docker/geostore-datasource-ovr.properties ${CATALINA_BASE}/conf; \
     fi
 
-# Add war files to be deployed
-COPY docker/*.war "${CATALINA_BASE}/webapps/"
+RUN apt-get update \
+    && apt-get install --yes postgresql-client-11 \
```
Comment (randomorder) [inline Dockerfile:39]: please split the PG client installation and packages cleanup in two distinct RUN steps and make the PostrgeSQL client version number an ENV defined at the beginning of the dockerfile so that it is easy to change it

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    restart: on-failure
+    healthcheck:
+      test: /usr/bin/pg_isready
+      interval: 5s
+      timeout: 10s
+      retries: 120
+    environment:
+      POSTGRES_PASSWORD: mysecretpassword
+    ports:
+      - "5432:5432"
```
Comment (randomorder) [inline docker-compose.yml:19]: please remove the port binding on the host if non needed. The end user may have a port clashin and postgresql is not secured

**Review by randomorder — CHANGES_REQUESTED** [review]
pleae check my comments

**Review by randomorder — COMMENTED** [review]
Any changes to the README regarding how to build / run the docker image @agpenton ?

**Comment by agpenton** [comment]
> README
Working on it

File: .gitignore [hunk 4]
```diff
@@ -9,6 +9,7 @@ coverage/
 web/client/libs/Cesium/
 web/cesium.zip
 web/.classpath
+web/overlays
```
Comment (randomorder) [inline .gitignore:12]: why this change?

File: docker/01-init-user.sh [hunk 5]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+set -e
+
+psql -v ON_ERROR_STOP=1 --username postgres --dbname postgres <<-EOSQL
```
Comment (randomorder) [inline docker/01-init-user.sh:4]: What about [geostore database initialization](https://mapstore.readthedocs.io/en/v2020.02.01/developer-guide/database-setup/)? Is it done automatically?

File: .gitignore (hunk 4 above)
Comment (randomorder) [reply .gitignore:12]: @agpenton ?

File: docker/01-init-user.sh (hunk 5 above)
Comment (agpenton) [reply docker/01-init-user.sh:4]: the Geostore database initialization was resolve here https://github.com/geosolutions-it/MapStore2/pull/6490/commits/93b4066b4ac0b464374b3e97225ec8921c510aaf

File: docker-compose.yml (hunk 3 above)
Comment (agpenton) [reply docker-compose.yml]: port mapping was removed in https://github.com/geosolutions-it/MapStore2/pull/6490/commits/c842a9ca18f196332990cbae86921c130de76141

File: Dockerfile (hunk 2 above)
Comment (agpenton) [reply Dockerfile]: PG client split and ENV added in this commit https://github.com/geosolutions-it/MapStore2/pull/6490/commits/81f252222042c06992b67e8602977de593b31564

File: Dockerfile (hunk 1 above)
Comment (agpenton) [reply Dockerfile]: fixed in this commit https://github.com/geosolutions-it/MapStore2/pull/6490/commits/81f252222042c06992b67e8602977de593b31564

File: .gitignore (hunk 4 above)
Comment (agpenton) [reply .gitignore:12]: nothing in particular, I just saw it new after I ran the environment, and since I didn't see it before I just ignored it.

**Comment by randomorder** [comment]
can you please take a look @lpasquali ?

**Comment by randomorder** [comment]
@lpasquali ?

**Comment by lpasquali** [comment]
Note, I had to upgrade docker-compose  from:

```
luca@vanth:~/Development/MapStore2$ docker-compose version
docker-compose version 1.26.2, build eefe0d31
docker-py version: 4.2.2
CPython version: 3.7.7
OpenSSL version: OpenSSL 1.1.0l  10 Sep 2019
luca@vanth:~/Development/MapStore2$ 
```
to docker-compose version: 

```
luca@vanth:~/Development/MapStore2$ docker-compose version
docker-compose version 1.28.5, build c4eb3a1f
docker-py version: 4.4.4
CPython version: 3.7.10
OpenSSL version: OpenSSL 1.1.0l  10 Sep 2019
luca@vanth:~/Development/MapStore2$ 
```
to start the composition

**Review by lpasquali — CHANGES_REQUESTED** [review]
For me it is working smootlhy maybe it would be good to notify that composition needs latest docker-compose binary in the readme

@agpenton @randomorder

File: Dockerfile [hunk 6]
```diff
@@ -1,22 +1,44 @@
...
 
 # Tomcat specific options
 ENV CATALINA_BASE "$CATALINA_HOME"
-ENV JAVA_OPTS="${JAVA_OPTS}  -Xms512m -Xmx512m -XX:MaxPermSize=128m"
+ENV MAPSTORE_WEBAPP_DST="${CATALINA_BASE}/webapps"
+ENV INITIAL_MEMORY="512m"
+ENV MAXIMUM_MEMORY="512m"
+ENV JAVA_OPTS="${JAVA_OPTS}  -Xms${INITIAL_MEMORY} -Xmx${MAXIMUM_MEMORY} -XX:MaxPermSize=128m"
+ENV GEOSTORE_OVR_OPT="-Dgeostore-ovr=${CATALINA_BASE}/conf/${OVR} \
+                      -Duser.timezone=UTC"
```
Comment (randomorder) [inline Dockerfile:18]: "-Duser.timezone=UTC" is not directly related with GEOSTORE

File: docker/01-init-user.sh (hunk 5 above)
Comment (randomorder) [reply docker/01-init-user.sh:4]: no, the database is empty

**Review by randomorder — CHANGES_REQUESTED** [review]

File: docker/01-init-user.sh (hunk 5 above)
Comment (randomorder) [reply docker/01-init-user.sh:4]: Database initialization script here:
https://github.com/geosolutions-it/geostore/blob/master/doc/sql/002_create_schema_postgres.sql

**Comment by tdipisa** [comment]
@randomorder I would like to include also GeoServer in the docker compose with the MS user integration. This is needed only as an option since in various cases only MS is needed (eg. existing SDI). This will allow users to setup in one shot a complete MS/GS SDI using Docker. In the future we can also consider adding also LDAP as an option for the docker compose.

**Review by tdipisa — COMMENTED** [review]
@randomorder I would like to include also GeoServer in the docker compose with the MS user integration. This is needed only as an option since in various cases only MS is needed (eg. existing SDIs). This will allow users to setup in one shot a complete MS/GS SDI using Docker. In the future we can also consider adding also LDAP as an option for the docker compose.

File: docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      POSTGRES_USER: postgres
+      POSTGRES_PASSWORD: postgres
+      POSTGRES_DB: postgres
+    volumes:
+    - pg_data:${PGDATA:-/var/lib/postgresql/data}:rw
+    - ./docker/01-init-user.sh:/docker-entrypoint-initdb.d/01-init-user.sh
+    ports:
+      - 5432
+    networks:
+      - intergration-test
```
Comment (randomorder) [inline docker-compose.yml:26]: I would rename the network to something more meaningful. Maybe "mapstore" ?

Comment (agpenton) [reply docker-compose.yml]: fixed here https://github.com/geosolutions-it/MapStore2/pull/6490/commits/f69942616b6d08cba28293e1cc20cda263a5dc22

File: .gitignore (hunk 4 above)
Comment (randomorder) [reply .gitignore:12]: please remove it

File: docker/01-init-user.sh (hunk 5 above)
Comment (agpenton) [reply docker/01-init-user.sh:4]: resolved here https://github.com/geosolutions-it/MapStore2/pull/6490/commits/a6c3879dbce8330ffe5431dddd1dc5d59735452f

File: Dockerfile (hunk 6 above)
Comment (agpenton) [reply Dockerfile]: resolved here https://github.com/geosolutions-it/MapStore2/pull/6490/commits/e597002a1188e6809cdeabaaca0129fb396e4121

File: .gitignore (hunk 4 above)
Comment (agpenton) [reply .gitignore]: resolved here https://github.com/geosolutions-it/MapStore2/pull/6490/commits/e597002a1188e6809cdeabaaca0129fb396e4121

File: docker/01-init-user.sh (hunk 5 above)
Comment (randomorder) [reply docker/01-init-user.sh:4]: please download the scripts from the geostore repo instead of copying the content

**Comment by agpenton** [comment]
@randomorder The PR is updated, I added a Dockerfile to build the Postgres and download the database initialization script, so always will be updated if the content change
